posix/console.c

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/corelibc/string.h>
+#include <openenclave/corelibc/unistd.h>
+#include <openenclave/enclave.h>
+#include <openenclave/internal/posix/device.h>
+#include <openenclave/internal/posix/fdtable.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/trace.h>
+#include "posix_t.h"
+
+#define DEVICE_NAME "consolefs"
+#define MAGIC 0x0b292bab
+
+typedef struct _file
+{
+    struct _oe_device base;
+    uint32_t magic;
+    oe_host_fd_t host_fd;
+} file_t;
+
+static file_t* _cast_file(const oe_device_t* device)
+{
+    file_t* file = (file_t*)device;
+
+    if (file == NULL || file->magic != MAGIC)
+        return NULL;
+
+    return file;
+}
+
+static int _consolefs_dup(oe_device_t* file_, oe_device_t** new_file_out)
+{
+    int ret = -1;
+    file_t* file = _cast_file(file_);
+    oe_host_fd_t retval = -1;
+
+    if (new_file_out)
+        *new_file_out = NULL;
+
+    if (!file || !new_file_out)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* Ask the host to perform this operation. */
+    {
+        oe_errno = 0;
+
+        if (oe_posix_dup_ocall(&retval, file->host_fd) != OE_OK)
+            OE_RAISE_ERRNO(OE_EINVAL);
+
+        if (retval == -1)
+            OE_RAISE_ERRNO(oe_errno);
+    }
+
+    /* Allocate and initialize a new file structure. */
+    {
+        file_t* new_file;
+
+        if (!(new_file = oe_calloc(1, sizeof(file_t))))
+            OE_RAISE_ERRNO(OE_ENOMEM);
+
+        memcpy(new_file, file, sizeof(file_t));
```

Comment:
mempcy_s

Reply:
Thank you Radhika. I fixed all of these in the PR.